### PR TITLE
[ci] Script to automatically clean up old Docker Hub commit tags

### PR DESCRIPTION
**Pull request by khluu**
## Why are these changes needed?
- The number of commit tags is accumulating on `ray` repo on Docker Hub, which needed to be clean
- This script automatically cleans up tags that are linked to images built at least 30 days ago.

### CODE REVIEW

**Comment by khluu** [comment]
@aslonnie  This is a draft script for you to review my logic on

File: ci/ray_ci/clean_up_old_docker_commit_tags.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import json
+from datetime import datetime, timezone, timedelta
+import re
+import subprocess
+from typing import Set
+
+
+def list_commit_shas():
+    """
+    Get list of commit SHAs on ray master branch from at least 30 days ago.
```
Comment (aslonnie) [inline ci/ray_ci/clean_up_old_docker_commit_tags.py:11]: you can just parse `git log` of the repo locally after cloned, which might be much easier (e.g. no need to do pages).

File: ci/ray_ci/clean_up_old_docker_commit_tags.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    token = get_auth_token_docker_hub("username", "password")
+    headers = {
+        "Authorization": f"Bearer {token}",
+    }
+    for tag in tags:
+        print(f"Deleting {tag}")  # TODO: delete this line
+        url = f"https://hub.docker.com/v2/repositories/{namespace}/{repository}/tags/{tag}"
+        response = requests.delete(url, headers=headers)
+        if response.status_code != 204:
+            print(f"Failed to delete {tag}")
```
Comment (aslonnie) [inline ci/ray_ci/clean_up_old_docker_commit_tags.py:130]: consider print the status code?

File: ci/ray_ci/clean_up_old_docker_commit_tags.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        )
+        if "errors" in response.json():
+            continue
+        result = response.json()["results"]
+        tags = [tag["name"] for tag in result]
+        # Check if tag is in list of commit SHAs
+        commit_tags = [
+            tag
+            for tag in tags
+            if len(tag.split("-")[0]) == 6 and tag.split("-")[0] in commit_short_shas
```
Comment (aslonnie) [inline ci/ray_ci/clean_up_old_docker_commit_tags.py:158]: you might want to be careful where there might be duplicates in `commit_short_shas`, that is, a recent commit (within 30 days) might share a same commit prefix with one from the past.

our guarding / criteria for delete or not should be something like this:

````
if it is recent (created / uploaded) within N days:
  do not delete
else if it has a commit prefix:
  if the commit prefix is within N days of recent commits:
    do not delete
  else:
    delele
````

File: ci/ray_ci/clean_up_old_docker_commit_tags.py (hunk 1 above)
Comment (khluu) [reply ci/ray_ci/clean_up_old_docker_commit_tags.py:10]: thank you! using `git log` is wayy easier

File: ci/ray_ci/clean_up_old_docker_commit_tags.py (hunk 2 above)
Comment (khluu) [reply ci/ray_ci/clean_up_old_docker_commit_tags.py]: done!

**Pull request closed without merging**